### PR TITLE
Implement checked exceptions 

### PR DESCRIPTION
**Pull request by JanMikes**
Hi, i am using exception rules for phpstan (https://github.com/pepakriz/phpstan-exception-rules) and i found out there is missing `@throws` annotation on `Uri::__construct()`. 

Because i am wrapping `new Uri()` with try+catch like this:
```php
try { new Uri(''); } 
catch (InvalidArgumentException | MalformedUriException) { /* .. */ }
```

I receive errors:
```
InvalidArgumentException is never thrown in the corresponding try block
GuzzleHttp\Psr7\Exception\MalformedUriException is never thrown in the corresponding try block
```

That is actually not true, because both of these exceptions are throw in constructor and both of these cases are already covered by tests:
- https://github.com/guzzle/psr7/blob/master/tests/UriTest.php#L108
- https://github.com/guzzle/psr7/blob/master/tests/UriTest.php#L139

I believe there will be some similar extension/rule for psalm or any other static analysis tools as well.

My proposal is to be transparent about which exceptions can be thrown and provide annotations to help IDE + tools understand the code better.

### CODE REVIEW

**Comment by JanMikes** [comment]
As well, i can prepare PR with adding exceptions extension to the CI and possibly fix all found issues, if you are up for it.

**Comment by GrahamCampbell** [comment]
Thanks. I thinks PR that adds throws doc to all the places would be a good idea. I usually follow the convention that anything that extends Error or RuntimeException is fair game and should not be documented, but everything else should.

**Comment by JanMikes** [comment]
> I usually follow the convention that anything that extends Error or RuntimeException is fair game and should not be documented

Hehe, seems like this convention differs 😄 

Source: https://github.com/pepakriz/phpstan-exception-rules#motivation
> These exceptions we called RuntimeException or maybe better "checked exception". All these exceptions should be checked. Therefore it must be either caught or written in @throws annotation. 

My opinion is that this convention may differ project per project and it is up to the developer whether they use LogicException or RuntimeException for that (or anything else). 

At the same time i believe, that for packages maintainers, especially for packages like guzzle that are used very widely, is the best way to document all exceptions and let the end user to choose which strategy they will use. WDYT? 


>  I thinks PR that adds throws doc to all the places would be a good idea

I will continue the work in this PR then, stay tuned 😉

**Comment by GrahamCampbell** [comment]
It's already a well established convention in Java to not document RuntimeException, and this is enforced by the compiler. I don't see a good reason to deviate from that.

**Comment by GrahamCampbell** [comment]
I disagree with what he wrote about LogicException too. Modern PHP code should use TypeError and ValueError to indicate code was incorrectly called: an actual bug in the caller code.

**Comment by JanMikes** [comment]
> to not document RuntimeException

Anyways we are "safe" here, because both `MalformedUriException` and `InvalidArgumentException` extends `LogicException` so following this rule, both should be documented via `@throws` annotation.

Im already preparing the code.

File: src/HttpFactory.php [hunk 1]
```diff
@@ -83,11 +95,17 @@ public function createServerRequest(string $method, $uri, array $serverParams =
...
+    /**
+     * @throws \InvalidArgumentException
+     */
     public function createResponse(int $code = 200, string $reasonPhrase = ''): ResponseInterface
     {
         return new Response($code, [], null, '1.1', $reasonPhrase);
     }
 
+    /**
+     * @throws \InvalidArgumentException
```
Comment (GrahamCampbell) [inline src/HttpFactory.php:107]: If this throws doc is not present on the psr interface, we are actually violating the contract here.

Comment (JanMikes) [reply src/HttpFactory.php:107]: I see. 2 possible solutions that come in my mind:
- Add `@throws` to the psr interface
- Do not add `@throws` here and add phpstan error to baseline so check passes.

Anyways as i would understand the contract, it says that `\Psr\Http\Message\RequestFactoryInterface::createRequest()` can never fail and will aways return instance of `RequestInterface`, which i believe is not true

Comment (GrahamCampbell) [reply src/HttpFactory.php:107]: I'm not saying your doc is wrong. :P As you say, either the psr interface is missing some information, or our implementation is bad. @Nyholm what do you think here?

File: phpstan-baseline.neon [hunk 2]
```diff
@@ -161,7 +161,7 @@ parameters:
 			path: src/Utils.php
 
 		-
-			message: "#^Parameter \\#2 \\$str of function fwrite expects string, string\\|false given\\.$#"
+			message: "#^Parameter \\#2 \\$[a-z]* of function fwrite expects string, string\\|false given\\.$#"
```
Comment (JanMikes) [inline phpstan-baseline.neon:164]: This was failing on php 8

File: phpstan-baseline.neon [hunk 3]
```diff
@@ -171,7 +171,16 @@ parameters:
 			path: src/Utils.php
 
 		-
-			message: "#^Unreachable statement \\- code above always terminates\\.$#"
+			message: "#^Strict comparison using === between '' and non-empty-string will always evaluate to false\\.$#"
```
Comment (JanMikes) [inline phpstan-baseline.neon:174]: Should be addressed in separated PR, not subject of this one

File: phpstan-baseline.neon [hunk 4]
```diff
@@ -171,7 +171,16 @@ parameters:
...
+
+		-
+			message: "#^Method GuzzleHttp\\\\Psr7\\\\LazyOpenStream::__get\\(\\) throws checked exception InvalidArgumentException but it's missing from the PHPDoc @throws tag.$#"
+			count: 1
+			path: src/StreamDecoratorTrait.php
 
+		-
+			message: "#^Method GuzzleHttp\\\\Psr7\\\\MultipartStream::__get\\(\\) throws checked exception InvalidArgumentException but it's missing from the PHPDoc @throws tag.$#"
+			count: 1
+			path: src/StreamDecoratorTrait.php
```
Comment (JanMikes) [inline phpstan-baseline.neon:186]: It was impossible to solve this one, it seems to be bug in phpstan, im already in touch with @ondrejmirtes about creating issue with bugreport

File: phpstan.neon.dist [hunk 5]
```diff
@@ -7,3 +7,11 @@ parameters:
...
     paths:
         - src
+    treatPhpDocTypesAsCertain: false
+    exceptions:
+        check:
+            missingCheckedExceptionInThrows: true
+            tooWideThrowType: true
+        uncheckedExceptionClasses:
+            - 'RuntimeException'
+            - 'BadMethodCallException'
```
Comment (JanMikes) [inline phpstan.neon.dist:17]: It is specific for this code:
```
protected function createStream(): StreamInterface
    {
        throw new \BadMethodCallException('Not implemented');
    }
```

I believe it is not goal to force developer to wrap this exception with try/catch

**Comment by JanMikes** [comment]
Hi, is there anything i could do to help this PR get merged? :-)

**Comment by GrahamCampbell** [comment]
not stale

File: .php_cs.cache [hunk 6]
```diff
@@ -0,0 +1 @@
...UriNormalizerTest.php":2702753674,"tests\/PumpStreamTest.php":810496326,"tests\/FnStreamTest.php":3023696406,"tests\/ReadSeekOnlyStream.php":2239940617,"tests\/StreamTest.php":3234337377,"tests\/AppendStreamTest.php":3613868310,"tests\/InflateStreamTest.php":2400745463,"tests\/ResponseTest.php":3210667136,"tests\/DroppingStreamTest.php":4096510809,"tests\/CachingStreamTest.php":3686162664,"tests\/Integration\/ServerRequestFromGlobalsTest.php":4090171619,"tests\/Integration\/server.php":819196779,"tests\/UriTest.php":1635324753,"tests\/UploadedFileTest.php":704018714,"tests\/MultipartStreamTest.php":3554036590,"tests\/RequestTest.php":101334953,"tests\/MimeTypeTest.php":1109422250,"tests\/MessageTest.php":650906072,"tests\/StreamDecoratorTraitTest.php":1056619207,"tests\/HasToString.php":472560092,"tests\/UriResolverTest.php":3187700976,"tests\/LimitStreamTest.php":561003037,"tests\/NoSeekStreamTest.php":3031719965,"tests\/BufferStreamTest.php":4089247971,"tests\/UtilsTest.php":1642749182,"tests\/QueryTest.php":1370167524,"tests\/LazyOpenStreamTest.php":1225196132,"tests\/ServerRequestTest.php":4098073084,"tests\/StreamWrapperTest.php":3770959201,"tests\/HeaderTest.php":3279788608}}
```
Comment (GrahamCampbell) [inline .php_cs.cache:1]: don't commit this

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: .php_cs.cache (hunk 6 above)
Comment (JanMikes) [reply .php_cs.cache]: Thank you, i missed this one during rebase.

**Comment by GrahamCampbell** [comment]
I'd rather change things in v3 to throw a `ValueError`, rather than changing out all the phpdoc to add a throws. Thanks for the PR though - it highlighted that we were violating the interfaces, albeit in edge cases.

**Comment by JanMikes** [comment]
Sure. I'm looking forward 😉

**Pull request closed without merging**